### PR TITLE
Add animal spawn setting.

### PR DESCRIPTION
**Pull request by Vasyan2006**
#### Summary
```SUMMARY: Features "Add animal spawn setting."```

#### Purpose of change
Zombies are main enemies and partially loot source while animals are main meat source.
 Increasing monster spawn will increase game difficulty but also increase food abundance, so separate spawn setting is required for better balancing.

#### Describe the solution
- Add new setting:
wildlife
- Add ``"is_animal" : true`` to some json monster groups, these groups will use new spawn rate. Other groups will use base scaling factor.
- Monster groups written in overmap_terrain now will use spawn scaling factor. This includes animals in forests and swamps.
- Add external option in game_balance.json to disable hardcoded monster clashes on swamps.

#### Additional context
There are 2 different code parts which choose monsters from the group, probably they must be merged somehow.

### CODE REVIEW

File: src/mapgen.cpp [hunk 1]
```diff
@@ -175,8 +175,26 @@ void map::generate( const int x, const int y, const int z, const time_point &whe
...
-        int pop = rng( spawns.population.min, spawns.population.max );
+
+    float spawn_density = 1.0f;
+    if( MonsterGroupManager::is_animal( spawns.group ) ) {
+        spawn_density = get_option< float >( "SPAWN_ANIMAL_DENSITY" );
+    } else {
+        spawn_density = get_option< float >( "SPAWN_DENSITY" );
+    }
+
+    // Apply a multipler to the number of monsters for really high densities.
```
Comment (Night-Pryanik) [inline src/mapgen.cpp]:
```suggestion
    // Apply a multiplier to the number of monsters for really high densities.
```

File: src/mapgen.cpp [hunk 2]
```diff
@@ -175,8 +175,26 @@ void map::generate( const int x, const int y, const int z, const time_point &whe
...
+    float odds_after_density = spawns.chance * spawn_density;
+    float max_odds = 100 - ( 100 - spawns.chance ) / 2;
```
Comment (Night-Pryanik) [inline src/mapgen.cpp]: Could be `const`.

File: src/mapgen.cpp [hunk 3]
```diff
@@ -175,8 +175,26 @@ void map::generate( const int x, const int y, const int z, const time_point &whe
...
+    float density_multiplier = 1.0f;
+    if( odds_after_density > max_odds ) {
+        density_multiplier = 1.0f * odds_after_density / max_odds;
+        odds_after_density = max_odds;
+    }
+    int spawn_count = roll_remainder( density_multiplier );
```
Comment (Night-Pryanik) [inline src/mapgen.cpp]: Could be `const`.

File: src/options.cpp [hunk 4]
```diff
@@ -1703,7 +1703,12 @@ void options_manager::add_options_world_default()
        );
 
     add( "SPAWN_DENSITY", "world_default", translate_marker( "Spawn rate scaling factor" ),
-         translate_marker( "A scaling factor that determines density of monster spawns." ),
+         translate_marker( "A scaling factor that determines density of monster spawns (except animals)." ),
```
Comment (Night-Pryanik) [inline src/options.cpp]:
```suggestion
         translate_marker( "A scaling factor that determines density of monster spawns (except for animals)." ),
```

File: data/json/game_balance.json [hunk 5]
```diff
@@ -145,5 +145,12 @@
     "info": "Enables CBM slots mechanics. If true CBM slots are enabled.",
     "stype": "bool",
     "value": false
+  },
+  {
+    "type": "EXTERNAL_OPTION",
+    "name": "DISABLE_SWAMP_CLASH",
+    "info": "Disable monster clashes on swamps.",
```
Comment (Night-Pryanik) [inline data/json/game_balance.json]: Add some more info on what "monster clashes on swamp" is.

File: src/options.cpp [hunk 6]
```diff
@@ -1703,7 +1703,12 @@ void options_manager::add_options_world_default()
...
+         0.0, 50.0, 1.0, 0.1
+       );
+
+    add( "SPAWN_ANIMAL_DENSITY", "world_default", translate_marker( "Animal spawn rate scaling factor" ),
+         translate_marker( "A scaling factor that determines density of wild, formerly domesticated and mutated animal spawns." ),
```
Comment (Night-Pryanik) [inline src/options.cpp]:
```suggestion
         translate_marker( "A scaling factor that determines density of wild, formerly domesticated and mutated animals spawns." ),
```

Comment (ifreund) [reply src/options.cpp]: That one's actually correct as a singular

Comment (Night-Pryanik) [reply src/options.cpp]: Enumeration of 1) wild, 2) formerly domesticated, and 3) mutated should be "animal"?

File: data/json/monstergroups.json [hunk 7]
```diff
@@ -1609,13 +1622,15 @@
...
       "monsters" : [
         { "monster" : "mon_duck", "freq" : 50, "cost_multiplier" : 0 }
       ]
   },{
       "type":"monstergroup",
       "name" : "GROUP_POND_FISH",
       "default" : "mon_null",
+      "is_animal" : true,
       "monsters" : [
         { "monster" : "mon_duck", "freq" : 50, "cost_multiplier" : 0 },
```
Comment (Night-Pryanik) [inline data/json/monstergroups.json]: I guess this is an error here: duck shouldn't belong to `GROUP_POND_FISH`.

File: src/options.cpp (hunk 6 above)
Comment (ifreund) [reply src/options.cpp]: "animal spawn" is kinda treated as one thing in this case so the plural would be "animal spawns."  Also it just sounds right to me. 

Either that or add an apostrophe: "animal's spawns" but that's a little awkward.

**Comment by kevingranade** [comment]
Please move this to a hidden option and add it as a mod. The options list
is already way too big and tweaking natural animal spawns isn't something
we should be suggesting that players should typically adjust.

**Comment by Vasyan2006** [comment]
And now I am getting 
``obj\Release\src\game.o:game.cpp|| undefined reference to `get_scaling_factor()'|`` 
compilation error. I will better not update this PR until 0.D will be released and all branches will return to usual routine.

**Comment by Vasyan2006** [comment]
Will reopen it shortly.

**Pull request closed without merging**